mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageDAO.java

```diff
@@ -223,58 +169,39 @@ private UDTValue toUDT(MessageAttachment messageAttachment) {
-                .setSet(USER_FLAGS, userFlagsSet(message))
-                .setLong(MOD_SEQ, message.getModSeq())
-                .setUUID(MESSAGE_ID, messageId.get())
-                .setLong(MOD_SEQ, oldModSeq))
-            .thenApply(optional -> optional
-                    .map(row -> row.getBool(CassandraConstants.LIGHTWEIGHT_TRANSACTION_APPLIED))
-                    .orElse(false));
-    }
-    
-    public CompletableFuture<Stream<Pair<MailboxMessage, Stream<MessageAttachmentById>>>> retrieveMessages(List<CassandraMessageId> messageIds, FetchType fetchType, Optional<Integer> limit) {
+    public CompletableFuture<Stream<Pair<MailboxMessage, Stream<MessageAttachmentById>>>> retrieveMessages(List<ComposedMessageIdWithMetaData> messageIds, FetchType fetchType, Optional<Integer> limit) {
         return retrieveRows(messageIds, fetchType, limit)
                 .thenApply(resultSet -> CassandraUtils.convertToStream(resultSet)
-                        .map(row -> message(row, fetchType)));
+                        .map(row -> message(row, messageIds, fetchType)));
     }
 
-    private CompletableFuture<ResultSet> retrieveRows(List<CassandraMessageId> messageIds, FetchType fetchType, Optional<Integer> limit) {
+    private CompletableFuture<ResultSet> retrieveRows(List<ComposedMessageIdWithMetaData> messageIds, FetchType fetchType, Optional<Integer> limit) {
         return cassandraAsyncExecutor.execute(
                 buildSelectQueryWithLimit(
                         buildQuery(messageIds, fetchType), 
                         limit)
                 );
     }
     
-    private Where buildQuery(List<CassandraMessageId> messageIds, FetchType fetchType) {
+    private Where buildQuery(List<ComposedMessageIdWithMetaData> messageIds, FetchType fetchType) {
         return select(retrieveFields(fetchType))
                 .from(TABLE_NAME)
                 .where(in(MESSAGE_ID, messageIds.stream()
+                        .map(ComposedMessageIdWithMetaData::getComposedMessageId)
+                        .map(ComposedMessageId::getMessageId)
+                        .map(messageId -> (CassandraMessageId) messageId)
                         .map(CassandraMessageId::get)
                         .collect(Collectors.toList())));
     }
 
-    private Pair<MailboxMessage, Stream<MessageAttachmentById>> message(Row row, FetchType fetchType) {
+    private Pair<MailboxMessage, Stream<MessageAttachmentById>> message(Row row, List<ComposedMessageIdWithMetaData> messageIds, FetchType fetchType) {
         try {
-            ComposedMessageId messageId = retrieveComposedMessageId(messageIdFactory.of(row.getUUID(MESSAGE_ID))).join();
+            ComposedMessageIdWithMetaData messageIdWithMetaData = retrieveComposedMessageId(messageIdFactory.of(row.getUUID(MESSAGE_ID)), messageIds);
+            ComposedMessageId messageId = messageIdWithMetaData.getComposedMessageId();
 
             SimpleMailboxMessage message =
```

Comment:
We should introduce a new type for this: it's a temporary instance as we don't yet have attachments.